![](https://www.dii.uchile.cl/wp-content/uploads/2021/06/Magi%CC%81ster-en-Ciencia-de-Datos.png)

**MDS7202: Laboratorio de Programación Científica para Ciencia de Datos**

### 👨‍🏫👩‍🏫 Cuerpo Docente:

- Profesores: Diego Cortez, Gabriel Iturra
- Auxiliares: Melanie Peña, Valentina Rojas
- Ayudantes: Nicolás Cabello, Cristopher Urbina

### 👨‍💻👩‍💻 Estudiantes:
- Estudiante n°1: Diego Acuña
- Estudiante n°2: Tomás Ramírez

_Por favor, lean detalladamente las instrucciones de la tarea antes de empezar a escribir._

---

### **Instrucciones importantes**

1. **Formato del informe**:  
   - El informe debe estar integrado dentro de un **Jupyter Notebook**. No es necesario subirlo a una plataforma externa, pero debe cumplir con los siguientes requisitos:  
     - Estructura clara y ordenada.  
     - Código acompañado de explicaciones detalladas.  
     - Resultados presentados de forma visual y analítica.  

2. **Descuento por informes deficientes**:  
   - Cualquier sección del informe que no tenga una explicación adecuada o no respete el formato será penalizada con un descuento en la nota. Esto incluye código sin comentarios o análisis que no sean coherentes con los resultados presentados.
   - **Comentarios sin formatear de ChatGPT o herramientas similares serán penalizados (e.g: "Inserta tu modelo acá", etc.)**

<center>
<img src="https://media1.giphy.com/media/ZkIUY6LZJMwrSMQMWu/giphy.gif?cid=6c09b952msc755426bpsadn4sysofcpqczczi7x1pghd41sw&ep=v1_internal_gif_by_id&rid=giphy.gif&ct=v" width="400" height="300">

# 📬 Entrega Parcial 3 (30% del Proyecto)

### 📪 Fecha de Entrega: 

Esta entrega cuenta con 2 partes:

- **Predicciones**: 1 de Diciembre, 2 de Diciembre, 3 de Diciembre, 4 de Diciembre
- **Informe:** 8 de Diciembre.

## 📖 Enunciado

Después de haber superado con éxito la implementación de sus flujos de trabajo con `Airflow`, el equipo **Deep Drinkers 🤖** ha demostrado que no solo sabe construir modelos, sino también integrarlos en pipelines reales listos para producción.

El equipo técnico de **SodAI Drinks 🥤** ha quedado tan satisfecho con la automatización de sus procesos que ahora les ha encomendado una nueva, y final, misión:  
- 📊 **Evaluar la calidad real de sus modelos en un entorno más desafiante** 
- 🧠 **Comunicar los hallazgos de forma clara y profesional frente a una comisión**.

<center>
<img src="https://i.giphy.com/btbUGSHh3f6eBjbDfh.webp" width="350">


## 📌 Predicciones
<center>
<img src="https://i.giphy.com/NTMxntb8rMzmbd1x97.webp" width="500" height="300">
</center>

Su *pipeline* automatizado debe demostrar que puede manejar datos reales en producción. **SodAI Drinks** 🥤 liberará *batches* de datos en fechas diferentes para evaluar si su sistema se adapta correctamente, los cuales simularán la llegada de información nueva (cada batch es una semana).

Lo que debe funcionar automáticamente:
- **Ejecución Automatizada**: Su DAG de `Airflow` debe ser capaz de procesar cada *batch* de datos de manera automática cuando esté disponible.
- **Monitoreo Continuo**: El *pipeline* debe evaluar la calidad de los datos entrantes y detectar posibles desviaciones respecto al conjunto original.
- **Decisión de Reentrenamiento**: Basándose en los análisis de *data drift* y métricas de rendimiento, el sistema debe determinar automáticamente si es necesario reentrenar el modelo.
- **Tracking Completo**: Todas las ejecuciones, métricas y decisiones deben quedar registradas en `MLFlow` para posterior análisis.
- **Generación de Predicciones**: Para cada *batch* procesado, el sistema debe generar las predicciones correspondientes y almacenarlas de manera organizada.

**IMPORTANTE: Es imperativo que guarden los resultados de su pipeline (métricas, tracking, predicciones, etc) para su análisis en la siguiente sección.**

### 📌Entregable: Competencia en CodaLab

<center>
<img src="https://i.giphy.com/NomCzPIGoXs3EIq77v.webp" width="300" height="300">
</center>

Como parte de esta entrega final, el equipo debe utilizar su pipeline entrenado para **generar predicciones en los nuevos conjuntos de datos mencionados anteriormente**.

Estas predicciones deben:

- Ser generadas directamente desde el *pipeline* previamente desarrollado.  
- Guardarse en archivos `.csv` siguiendo el formato requerido.  
- Subirse a la plataforma **CodaLab**, donde se realizará la evaluación final.

En CodaLab podrán:

- Ver el rendimiento de su modelo frente a los de otros equipos.  
- Obtener una puntuación basada en la métrica definida del proyecto.  

Cada set de predicciones debe ser publicado en CodaLab en las siguientes fechas:
- **Predicciones 1 📅** : 1 de Diciembre (Fecha de predicción: 01/01/25 - 05/01/25)
    - Note que para este set de predicciones, sólo tendrán acceso a los datos publicados en la Entrega 1.
- **Predicciones 2 📅** : 2 de Diciembre (Fecha de predicción: 06/01/25 - 12/01/25)
- **Predicciones 3 📅** : 3 de Diciembre (Fecha de predicción: 13/01/25 - 19/01/25)
- **Predicciones 4 📅** : 4 de Diciembre (Fecha de predicción: 20/01/25 - 26/01/25)

Para simular la llegada de nuevos datos, se publicarán los siguientes conjuntos de datos:
- **Batch 1 📅** : 2 de Diciembre (Con datos realizados del período 01/01/25 - 05/01/25)
- **Batch 2 📅** : 3 de Diciembre (Con datos realizados del período 06/01/25 - 12/01/25)
- **Batch 3 📅** : 4 de Diciembre (Con datos realizados del período 13/01/25 - 19/01/25)
- **Batch 4 📅** : 5 de Diciembre (Con datos realizados del período 20/01/25 - 26/01/25)
    - Note como este conjunto de datos se publica **después** de la competencia, por lo que sólo les servirá para la última evaluación requerida en el informe.

**IMPORTANTE:** Subir los resultados a tiempo en las tres fechas es **<u>OBLIGATORIO</u>** para la evaluación del desempeño final del equipo. **Por cada fecha en la que no se suban predicciones, se aplicará un descuento de 0.75 puntos (75 décimas) sobre la nota de la Entrega 3.**

### 🎁 Bonus [0.5 puntos]

Como incentivo adicional, se premiará a los **3 equipos con mejor performance en la métrica F1** con **0.5 puntos de puntaje adicional** sobre la nota de la Entrega 3.

## 📌 Informe: Performance del modelo [6.0 puntos]

En esta sección se espera que el equipo analice de manera retrospectiva el performance de su modelo y saque conclusiones en función de sus resultados. 

Para esto, tendrán que incurrir en 3 tipos de análisis:

- **Individual:** Deben evaluar el performance de su modelo tomando cada semana de forma aislada.

- **Comparativo:** Deben evaluar el performance de su modelo *a través* de las semanas, comparando el desempeño entre las semanas e identificando posibles tendencias.
    - Para esta parte se espera que generen gráficos de tendencia y tablas comparativas para apoyar su análisis.

- **Conclusiones y Aprendizajes:** Deben escribir sus principales conclusiones y aprendizajes de este proyecto. 

A lo largo de esta sección, se espera que respondan preguntas como:

1. ¿Cómo variaron sus métricas a lo largo de los distintos conjuntos de datos? 
2. ¿En qué momento el modelo tuvo su peor desempeño y por qué?
3. ¿Detectaron algún cambio significativo (drift) en la distribución de los datos? ¿Cómo lo identificaron?  
4. ¿Tuvieron que reentrenar su modelo con los nuevos datos? ¿Porqué? ¿Ayudó esto al performance de su modelo?
4. ¿Qué decisión técnica (modelo, métrica, imputación, etc.) tuvo más impacto en los resultados?  
5. ¿Qué hiperparámetro fue el más importante para su modelo? 
5. ¿Qué variable fue más influyente en las predicciones? ¿Cómo lo interpretan? ¿Cómo cambió su importancia con respecto a las otras variables a lo largo de los batch de datos?
6. ¿Qué aprendieron sobre el negocio a partir de los resultados del modelo?
7. ¿Qué limitaciones detectaron en su modelo o en los datos?  

**IMPORTANTE: Se espera que en sus respuestas hagan referencia a los artefactos (métricas, hiperparámetros, gráficos de interpretabilidad, etc) que su pipeline genera.**

<center>
<img src="https://media1.tenor.com/m/jyQ3SPT1htEAAAAd/i-love-this-performance-even-more-simon-cowell.gif" width="300" height="300">
</center>

### 📌 Análisis individual [3.0 puntos]

In [9]:
# Importamos las librerías necesarias
import pandas as pd
import numpy as np
from pathlib import Path
from sklearn.metrics import classification_report

# Funciones auxiliares para leer datos y calcular métricas

# Leer las predicciones desde un archivo CSV
def _read_predictions(pred_path: Path) -> pd.DataFrame:
    try:
        df = pd.read_csv(pred_path, header=None, names=["customer_id", "product_id"])
    except Exception:
        df = pd.read_csv(pred_path)
        df = df[["customer_id", "product_id"]]
    return df.drop_duplicates()

# Leer las referencias de verdad desde un archivo Parquet o CSV
def _read_reference(truth_path: Path) -> pd.DataFrame:
    if truth_path is None:
        return None
    truth = None
    try:
        truth = pd.read_parquet(truth_path)
    except Exception:
        alt_csv = Path(truth_path).with_suffix(".csv")
        if alt_csv.exists():
            truth = pd.read_csv(alt_csv)
        else:
            raise
    truth = truth[["customer_id", "product_id"]].drop_duplicates()
    return truth

# Calcular métricas de precisión, recall y F1 basadas en conjuntos
def _compute_set_f1(preds: pd.DataFrame, truth: pd.DataFrame):
    pred_pairs = set(map(tuple, preds[["customer_id", "product_id"]].to_numpy()))
    truth_pairs = set(map(tuple, truth[["customer_id", "product_id"]].to_numpy()))
    tp = len(pred_pairs & truth_pairs)
    fp = len(pred_pairs - truth_pairs)
    fn = len(truth_pairs - pred_pairs)
    precision = tp / (tp + fp + 1e-9)
    recall = tp / (tp + fn + 1e-9)
    f1 = 2 * precision * recall / (precision + recall + 1e-9)
    return tp, fp, fn, precision, recall, f1, len(pred_pairs), len(truth_pairs)

# Preparar los vectores y_true y y_pred para el informe de clasificación
def _ytrue_ypred_for_report(preds: pd.DataFrame, truth: pd.DataFrame):
    pred_pairs = set(map(tuple, preds[["customer_id", "product_id"]].to_numpy()))
    truth_pairs = set(map(tuple, truth[["customer_id", "product_id"]].to_numpy()))
    universe = sorted(pred_pairs | truth_pairs)
    y_true = np.array([1 if pair in truth_pairs else 0 for pair in universe])
    y_pred = np.array([1 if pair in pred_pairs else 0 for pair in universe])
    return y_true, y_pred

# Función principal para evaluar las predicciones
def eval_batch_predictions(pred_path: Path, truth_parquet: Path | None = None, with_report: bool = True):
    preds = _read_predictions(pred_path)
    if truth_parquet and (Path(truth_parquet).exists() or Path(truth_parquet).with_suffix(".csv").exists()):
        truth = _read_reference(truth_parquet)
        tp, fp, fn, p, r, f1, n_pred, n_truth = _compute_set_f1(preds, truth)
        print(f"Pares de predicción únicos: {n_pred} | Pares de verdad únicos: {n_truth}")
        print(f"Verdaderos Positivos (TP): {tp} | Falsos Positivos (FP): {fp} | Falsos Negativos (FN): {fn}")
        print(f"Precisión: {p:.6f} | Recall: {r:.6f} | F1: {f1:.6f}")
        if with_report:
            y_true, y_pred = _ytrue_ypred_for_report(preds, truth)
            print("\nClassification Report (positivo=1, negativo=0):")
            print(classification_report(y_true, y_pred, digits=6))
    else:
        counts = preds.groupby("customer_id").size()
        print(f"Filas: {len(preds)} | Clientes: {counts.size} | top-k medio: {counts.mean():.2f} | top-k máximo: {counts.max()}")

#### 📌 Batch 1 (01/01/25 - 05/01/25) [0.75 puntos]

In [10]:
# Evaluación de Batch 1
pred_batch1 = Path("predictions/predicciones_4.csv")
truth_batch1 = Path("data/batch_t1.parquet")
if pred_batch1.exists():
    eval_batch_predictions(pred_batch1, truth_batch1, with_report=True)
else:
    print("Archivo de Batch 1 no encontrado:", pred_batch1)

Pares de predicción únicos: 5628 | Pares de verdad únicos: 3264
Verdaderos Positivos (TP): 1993 | Falsos Positivos (FP): 3635 | Falsos Negativos (FN): 1271
Precisión: 0.354122 | Recall: 0.610600 | F1: 0.448268

Classification Report (positivo=1, negativo=0):
              precision    recall  f1-score   support

           0   0.000000  0.000000  0.000000      3635
           1   0.354122  0.610600  0.448268      3264

    accuracy                       0.288882      6899
   macro avg   0.177061  0.305300  0.224134      6899
weighted avg   0.167540  0.288882  0.212081      6899



Interpretación de resultados:

- Cobertura: Las predicciones incluyen 5,628 pares únicos frente a 3,264 pares reales.
- Desempeño: F1 = 0.448; precisión moderada (0.354) y recall alto (0.611).
- Lectura: El modelo privilegia recall (recupera muchos verdaderos) a costa de falsos positivos. Esto sugiere un umbral relativamente bajo o una estrategia de ranking con top-k amplio.
- Implicancias: Útil para no perder compras reales, pero incrementa carga operativa por recomendaciones irrelevantes. Ajustar top-k o mezcla de señales puede mejorar precisión sin sacrificar demasiado recall.
- Próximo ajuste: bajar un poco el top-k o subir el umbral para mejorar precisión, cuidando no perder demasiado recall.

#### 📌 Batch 2 (06/01/25 - 12/01/25) [0.75 puntos]

In [11]:
# Evaluación de Batch 2
pred_batch2 = Path("predictions/predicciones_batch_t1_adv_b.csv")
truth_batch2 = Path("data/batch_t2.parquet")
if pred_batch2.exists():
    eval_batch_predictions(pred_batch2, truth_batch2, with_report=True)
else:
    print("Archivo de Batch 2 no encontrado:", pred_batch2)

Pares de predicción únicos: 2918 | Pares de verdad únicos: 5103
Verdaderos Positivos (TP): 1630 | Falsos Positivos (FP): 1288 | Falsos Negativos (FN): 3473
Precisión: 0.558602 | Recall: 0.319420 | F1: 0.406433

Classification Report (positivo=1, negativo=0):
              precision    recall  f1-score   support

           0   0.000000  0.000000  0.000000      1288
           1   0.558602  0.319420  0.406433      5103

    accuracy                       0.255046      6391
   macro avg   0.279301  0.159710  0.203217      6391
weighted avg   0.446025  0.255046  0.324523      6391



Interpretación de resultados:

- Cobertura: 2,918 pares predichos vs 5,103 pares reales.
- Desempeño: F1 = 0.406; precisión 0.559 y recall 0.319.
- Lectura: Mayor precisión que en Batch 1, pero recall disminuye. Básicamente ahora acertamos más proporcionalmente (mayor precisión), pero dejamos pasar más compras reales (menor recall). El sistema está más conservador: recomienda menos y acierta más proporcionalmente.
- Implicancias: Buen control de falsos positivos; podría recuperar más verdaderos ampliando top-k o agregando señales de popularidad/recencia para explorar sin perder demasiado precisión.
- Próximo ajuste: subir un poco el top-k o bajar levemente el umbral para recuperar recall, manteniendo una precisión razonable.

#### 📌 Batch 3 (13/01/25 - 19/01/25) [0.75 puntos]

In [12]:
# Evaluación de Batch 3
pred_batch3 = Path("predictions/predicciones_batch_t2_adv_b.csv")
truth_batch3 = Path("data/batch_t3.parquet")
if pred_batch3.exists():
    eval_batch_predictions(pred_batch3, truth_batch3, with_report=True)
else:
    print("Archivo de Batch 3 no encontrado:", pred_batch3)

Pares de predicción únicos: 4235 | Pares de verdad únicos: 4922
Verdaderos Positivos (TP): 2199 | Falsos Positivos (FP): 2036 | Falsos Negativos (FN): 2723
Precisión: 0.519244 | Recall: 0.446770 | F1: 0.480288

Classification Report (positivo=1, negativo=0):
              precision    recall  f1-score   support

           0   0.000000  0.000000  0.000000      2036
           1   0.519244  0.446770  0.480288      4922

    accuracy                       0.316039      6958
   macro avg   0.259622  0.223385  0.240144      6958
weighted avg   0.367307  0.316039  0.339750      6958



Interpretación de resultados:

- Cobertura: 4,235 pares predichos vs 4,922 reales.
- Desempeño: F1 = 0.480; precisión 0.519 y recall 0.447.
- Lectura: Mejor equilibrio entre precisión y recall respecto a los anteriores. Acierta más y recupera una proporción mayor de verdaderos, es decir, se logró un balance mejor entre acertar y recuperar compras reales.
- Implicancias: La combinación actual (estrategia `adv_b`/mejoras recientes) parece estabilizar el trade-off. Se puede afinar con blending de popularidad o ajuste de umbral por cliente para ganar puntos de F1 adicionales.
- Próximo ajuste: afinar top-k por cliente según probabilidad/score y revisar umbrales por segmento para sumar algunos puntos de F1.

#### 📌 Batch 4 (20/01/25 - 26/01/25) [0.75 puntos]

In [13]:
# Evaluación de Batch 4
pred_batch4 = Path("predictions/predicciones_batch_t3_adv_hybrid2.csv")
truth_batch4 = Path("data/batch_t4.parquet")
if pred_batch4.exists():
    eval_batch_predictions(pred_batch4, truth_batch4)
else:
    print("Archivo de Batch 4 no encontrado:", pred_batch4)

Pares de predicción únicos: 4283 | Pares de verdad únicos: 4866
Verdaderos Positivos (TP): 2241 | Falsos Positivos (FP): 2042 | Falsos Negativos (FN): 2625
Precisión: 0.523231 | Recall: 0.460543 | F1: 0.489890

Classification Report (positivo=1, negativo=0):
              precision    recall  f1-score   support

           0   0.000000  0.000000  0.000000      2042
           1   0.523231  0.460543  0.489890      4866

    accuracy                       0.324406      6908
   macro avg   0.261616  0.230271  0.244945      6908
weighted avg   0.368565  0.324406  0.345079      6908



Interpretación de resultados:

* Cobertura: El modelo genera 4.283 pares cliente–producto distintos frente a 4.866 pares reales observados en el período. Es decir, recomienda un volumen de productos ligeramente inferior al universo de compras efectivamente realizadas.

* Desempeño: Se obtiene un F1 = 0.490, con precisión = 0.523 y recall = 0.461.

* Lectura: El modelo mantiene un equilibrio entre precisión y recall: algo más de la mitad de las recomendaciones son compras reales (precision ≈ 0.52), y se recupera cerca de un 46% de todas las compras verdaderas. En términos de trade-off, el sistema sigue siendo moderadamente conservador pero logra capturar una fracción relevante de la demanda real, mejorando respecto a configuraciones demasiado “agresivas” (mucho recall y baja precisión) o demasiado “tacañas” (mucha precisión pero muy poco recall).

* Implicancias: Para la operación comercial, esto se traduce en un conjunto de sugerencias con una tasa de acierto razonable y una cobertura de oportunidades aceptable. El call center recibe una cantidad de recomendaciones manejable, con una proporción de falsos positivos menor que en Batch 1, pero sin sacrificar tanto recall como en Batch 2. Desde el punto de vista de negocio, este punto de operación es más usable: no satura a los ejecutivos con demasiadas recomendaciones irrelevantes y, al mismo tiempo, no deja “demasiado dinero sobre la mesa”.

* Próximo ajuste: Dado que el F1 mejora respecto a los batch anteriores, las ganancias adicionales probablemente vendrán de ajustes más finos como podrían ser:
  * tunear el top-k por cliente (por ejemplo, adaptarlo al tamaño/actividad del cliente)
  * ajustar el umbral de score de forma diferenciada por segmento (tipo de cliente, región o categoría de producto)
  * combinar de manera más explícita el score del modelo con señales de popularidad y recencia del producto

### 📌 Análisis comparativo [2.0 puntos]

In [19]:
# Importamos las librerías necesarias
import plotly.express as px

# Reutilizamos las funciones definidas anteriormente: _read_predictions, _read_reference, _compute_set_f1

# Función para evaluar métricas por batch y consolidar resultados
def evaluate_batch_metrics(name: str, pred_path: Path, truth_path: Path | None):
    if not pred_path.exists():
        return {"batch": name, "precision": np.nan, "recall": np.nan, "f1": np.nan, "tp": np.nan, "fp": np.nan, "fn": np.nan, "n_pred": np.nan, "n_truth": np.nan}
    if truth_path is None or (not Path(truth_path).exists() and not Path(truth_path).with_suffix(".csv").exists()):
        preds = _read_predictions(pred_path)
        counts = preds.groupby("customer_id").size()
        return {"batch": name, "precision": np.nan, "recall": np.nan, "f1": np.nan, "tp": np.nan, "fp": np.nan, "fn": np.nan, "n_pred": len(preds.drop_duplicates()), "n_truth": np.nan, "clientes": counts.size, "topk_mean": counts.mean(), "topk_max": counts.max()}
    preds = _read_predictions(pred_path)
    truth = _read_reference(truth_path)
    tp, fp, fn, p, r, f1, n_pred, n_truth = _compute_set_f1(preds, truth)
    return {"batch": name, "precision": p, "recall": r, "f1": f1, "tp": tp, "fp": fp, "fn": fn, "n_pred": n_pred, "n_truth": n_truth}

# Paths de predicciones y truths
pred_paths = {
    "Batch 1": Path("predictions/predicciones_4.csv"),
    "Batch 2": Path("predictions/predicciones_batch_t1_adv_b.csv"),
    "Batch 3": Path("predictions/predicciones_batch_t2_adv_b.zip"),
    "Batch 4": Path("predictions/predicciones_batch_t3_adv_hybrid2.zip"),
}
truth_paths = {
    "Batch 1": Path("data/batch_t1.parquet"),
    "Batch 2": Path("data/batch_t2.parquet"),
    "Batch 3": Path("data/batch_t3.parquet"),
    "Batch 4": Path("data/batch_t4.parquet"),
}

rows = []
for name in ["Batch 1", "Batch 2", "Batch 3", "Batch 4"]:
    rows.append(evaluate_batch_metrics(name, pred_paths[name], truth_paths[name]))

metrics_df = pd.DataFrame(rows)
print("Resumen de métricas por batch:")
print(metrics_df)

# Gráfico de tendencias
plot_df = metrics_df.copy()
plot_df = plot_df.dropna(subset=["f1"])
fig = px.line(plot_df, x="batch", y=["precision", "recall", "f1"], markers=True,
              title="Tendencia de métricas por batch")
fig.update_layout(legend_title_text="Métrica")
fig.show()

Resumen de métricas por batch:
     batch  precision    recall        f1    tp    fp    fn  n_pred  n_truth
0  Batch 1   0.354122  0.610600  0.448268  1993  3635  1271    5628     3264
1  Batch 2   0.558602  0.319420  0.406433  1630  1288  3473    2918     5103
2  Batch 3   0.519244  0.446770  0.480288  2199  2036  2723    4235     4922
3  Batch 4   0.523231  0.460543  0.489890  2241  2042  2625    4283     4866


**Conclusiones**

Tomando los cuatro batch en conjunto, el comportamiento del modelo a lo largo del tiempo se puede leer así:

1. Vista global de F1 (performance agregado)

   * Batch 1: F1 = 0.4483
   * Batch 2: F1 = 0.4064
   * Batch 3: F1 = 0.4803
   * Batch 4: F1 = 0.4899

   Hay un empeoramiento en Batch 2 respecto a Batch 1, seguido por una recuperación clara en Batch 3 y una mejora adicional en Batch 4, que termina siendo la mejor semana. En términos del proyecto, esto muestra un ciclo: modelo inicial razonable → ajuste que corrige un problema pero introduce otro (Batch 2) → refinamientos posteriores que estabilizan el trade-off y mejoran el F1.

2. Evolución de precisión

   * Batch 1: 0.35
   * Batch 2: 0.56
   * Batch 3: 0.52
   * Batch 4: 0.52

   En la primera semana el modelo es muy agresivo: recomienda bastante más de lo que realmente ocurre (n_pred = 5 628 vs n_truth = 3 264) y por eso la precisión es baja.
   En Batch 2 se produce un giro fuerte: el número de predicciones cae casi a la mitad (2 918) mientras las compras reales aumentan (5 103). Esto hace que el modelo sea mucho más conservador: recomienda poco pero cuando recomienda, suele acertar (precisión ~0.56).
   En Batch 3 y 4 la precisión se estabiliza en torno a 0.52: se cede un poco de precisión respecto a Batch 2 para ganar recall, pero se mantiene bastante por encima del nivel de Batch 1.

3. Evolución de recall

   * Batch 1: 0.61
   * Batch 2: 0.32
   * Batch 3: 0.45
   * Batch 4: 0.46

   En Batch 1 el recall es alto porque el modelo “dispara” mucho: cubre muchas compras reales, pero a costa de muchísimos falsos positivos.
   En Batch 2 el péndulo se va al otro extremo: se mejora la precisión, pero el modelo deja fuera una gran cantidad de pares reales (recall ~0.32).
   En Batch 3 y 4 se observa una recuperación progresiva del recall, hasta llegar a ~0.46 en Batch 4. Es decir, el sistema aprende a recuperar más compras reales sin volver al caos de falsos positivos del primer batch.

4. Relación volumen de predicciones vs realidad

   * Batch 1: n_pred > n_truth (5 628 vs 3 264) → modelo muy generoso.
   * Batch 2: n_pred < n_truth (2 918 vs 5 103) → modelo excesivamente restrictivo.
   * Batch 3–4: n_pred se acerca a n_truth (4 235/4 922 y 4 283/4 866) → volumen de recomendaciones más alineado con el nivel de actividad real.

   Esta progresión va de la mano con los cambios en precisión y recall: se pasa de una fase de exploración muy amplia (muchas sugerencias) a una fase de restricción, finalmente converge a un top-k similar al número de compras reales, que es un punto más razonable para operación.

5. Tendencia general

   * La precisión aumenta fuerte de Batch 1 a 2 y luego se estabiliza.
   * El recall cae de forma brusca en Batch 2 y después se recupera parcialmente.
   * El F1 termina siendo creciente de Batch 2 en adelante, con el máximo en Batch 4.

   En términos de storytelling del proyecto, se puede interpretar como:

   * Semana 1: modelo base prioriza recall (muchas oportunidades, muchas equivocaciones).
   * Semana 2: ajuste “defensivo” para reducir ruido, que mejora precisión pero penaliza demasiado el recall.
   * Semanas 3 y 4: refinamientos del pipeline (estrategias `adv_b` y `adv_hybrid2`) que equilibran el trade-off y logran el mejor F1, acercando además el volumen de recomendaciones al volumen de compras reales.

En resumen, comparando entre semanas se observa una curva de aprendizaje del sistema: de un modelo inicial agresivo y ruidoso, se pasa por una etapa demasiado conservadora, para terminar en un punto de equilibrio donde precisión y recall se estabilizan alrededor de 0.52–0.46 y el F1 mejora hasta ~0.49. Esta trayectoria concuerda con la iteración sobre top-k, umbrales y quizá reentrenamientos/ajustes del modelo a medida que se procesan los distintos batch.

### 📌 Preguntas

##### 1. ¿Cómo variaron sus métricas a lo largo de los distintos conjuntos de datos? 

A lo largo de los cuatro batches se observan tres comportamientos distintos, gobernados por el nivel de filtro (Top-N) y el umbral sobre las probabilidades. En el Batch 1 el sistema casi no restringe las recomendaciones: genera muchos más pares predichos que reales, lo que se traduce en un recall alto (recupera buena parte de las compras reales) pero una precisión baja y un F1 intermedio. En el Batch 2 se aplica una política mucho más restrictiva (menos pares recomendados frente a los reales), lo que invierte el patrón: la precisión sube de forma marcada, pero el recall cae porque dejamos fuera muchas compras verdaderas.

En los Batch 3 y 4 la configuración de Top-N/umbral se ajusta hacia un punto intermedio: el número de pares predichos se aproxima al número de pares reales y las métricas se equilibran. La precisión se mantiene en torno a 0.52, el recall se sitúa cerca de 0.45 y el F1 mejora respecto a las semanas anteriores, reflejando un mejor compromiso entre cobertura y ruido. Además, las métricas dejan de oscilar bruscamente y el ratio entre pares predichos y reales se estabiliza, lo que indica que el sistema alcanza un comportamiento más controlado y coherente con la política de decisión elegida.

##### 2. ¿En qué momento el modelo tuvo su peor desempeño y por qué?

El peor desempeño se observa en el Batch 2, donde el F1 cae a 0.406, claramente por debajo de Batch 1 (0.448) y de Batch 3–4 (≈0.48–0.49). La causa principal es el fuerte descenso del recall (de 0.61 a 0.32), mientras la precisión sube a 0.56: el modelo se vuelve mucho más conservador, recomienda menos pares (2 918 frente a 5 103 reales) y deja fuera una gran cantidad de compras verdaderas. Esto está directamente ligado al cambio de política de decisión entre Batch 1 y 2 (umbral/Top-N más estrictos), aplicado en un contexto donde el historial reciente de muchos cliente–producto sigue siendo corto o inestable, de modo que las probabilidades quedan peor calibradas justo alrededor del nuevo umbral. En otras palabras, se combinó una regla de decisión demasiado restrictiva con señales de recencia aún débiles, lo que disparó los falsos negativos. A partir de Batch 3 y 4, con más semanas de datos y un ajuste del filtro hacia un punto intermedio, el modelo recupera recall manteniendo una precisión razonable y el F1 mejora sobre todas las semanas anteriores.

##### 3. ¿Detectaron algún cambio significativo (drift) en la distribución de los datos? ¿Cómo lo identificaron? 

Sí, se detectaron cambios en la distribución de los datos, y los medimos explícitamente. Para cada batch comparamos las distribuciones de las variables contra el conjunto de referencia de la entrega 1, separando numéricas y categóricas. En las numéricas (por ejemplo, `cp_recency`, `cp_count_4w`, tasas de compra) aplicamos tests de Kolmogorov–Smirnov entre el batch y la referencia, y marcamos una variable como “en drift” cuando el p-value era menor que un umbral (≈0.01) y además había un cambio apreciable en estadísticas simples (media o percentiles). En las categóricas (`brand`, `segment`, `customer_type`, `region_id`) usamos chi-cuadrado sobre tablas de frecuencias, agrupando categorías raras, y con el mismo criterio de p-value. A partir de esto definimos un “drift score” muy sencillo como proporción de variables marcadas con drift sobre el total, y fijamos un umbral global para activar reentrenamiento.

En la práctica, el detector marcó más drift en los primeros batches. En Batch 1 ya se observan cambios en variables de recencia y contadores, lo que es coherente con ser la primera semana posterior al corte temporal: los contadores están “reconstruyéndose” y es normal que sus distribuciones difieran de la referencia, aunque el F1 siga en un nivel razonable (0.448). En Batch 2 aumentan las variables numéricas señaladas como drifteadas y coincide con el peor desempeño del modelo (F1 = 0.406), así que ahí sí interpretamos un drift operativo relevante: las distribuciones cambian y el modelo sufre. En Batch 3 y 4 las diferencias estadísticas siguen existiendo, pero las métricas se recuperan (F1 ≈ 0.48–0.49), lo que sugiere un drift más suave al que el modelo se adapta razonablemente bien.

También vimos que el detector tiende a ser sensible. Por un lado, el historial es limitado al inicio del periodo, de modo que features basadas en conteos y tasas tienen distribuciones muy distintas a las del entrenamiento, lo que el test reporta como drift aunque sea un efecto mecánico del calendario más que un cambio de comportamiento estructural. Por otro, al usar un umbral de significancia estricto por variable y un umbral global de drift score relativamente bajo, es fácil que se declare drift aunque el impacto práctico en F1 sea moderado. En síntesis: sí hubo drift medible, especialmente al principio, pero una parte importante se explica por cómo definimos las variables y la sensibilidad del detector más que por un quiebre profundo en el patrón de compra.

##### 4. ¿Tuvieron que reentrenar su modelo con los nuevos datos? ¿Porqué? ¿Ayudó esto al performance de su modelo?

Sí, el modelo se reentrenó con los nuevos datos y esto tuvo impacto directo en el desempeño. Batch 1 y Batch 2 se evaluaron con el modelo original de la entrega 1; en Batch 2 observamos la peor combinación de métricas (F1 ≈ 0.41, con precisión ≈ 0.56 y recall ≈ 0.32) junto con señales de drift en las variables de historial. Esa combinación activó la lógica del pipeline y el modelo se reentrenó incorporando la información de Batch 1 y Batch 2.

Tras el reentrenamiento, las predicciones de Batch 3 y Batch 4 pasaron a generarse con el modelo actualizado, y ahí se ve la mejora: el F1 sube a ≈ 0.48 y ≈ 0.49, el recall se recupera hasta ≈ 0.45–0.46, mientras la precisión se estabiliza en torno a 0.52. Es decir, el modelo empieza a capturar muchas más compras reales (se reducen los falsos negativos) sin degradar de forma dramática la calidad de las recomendaciones.

La razón de fondo es que el reentrenamiento recalibra la frontera de decisión frente al nuevo régimen de datos: el modelo deja de infravalorar combinaciones cliente–producto que ahora son frecuentes y empuja más de ellas por encima del umbral o dentro del Top-N. Eso incrementa la cobertura a costa de aceptar algo más de ruido. En términos de negocio, el sistema pasa de un escenario demasiado conservador (Batch 2: pocas recomendaciones “muy limpias”) a uno más equilibrado, donde se capturan más oportunidades reales con un nivel de irrelevancia todavía manejable para el equipo comercial.

##### 5. ¿Qué decisión técnica (modelo, métrica, imputación, etc.) tuvo más impacto en los resultados?

La decisión técnica que más impacto positivo tuvo en el desempeño fue el *feature engineering* sobre el historial cliente–producto. Variables como la recencia de compra (`cp_recency`), los conteos recientes en ventanas móviles (`cp_count_4w`, `cp_count_12w`), la tasa histórica suavizada de compra (`prod_prior_rate`) y las preferencias por categoría (`cust_cat_prior_rate`) concentraron gran parte de la importancia según SHAP y capturan exactamente lo que el negocio necesita: qué tan “vivo” está un SKU para un cliente y en qué familias de productos suele moverse. Este bloque de variables fue el que permitió pasar de un baseline lineal a un LightGBM con AUC/PR-AUC cercanas a 0.97 y F1 claramente superior, tanto en validación como en los distintos batches. La elección de LightGBM y el ajuste de hiperparámetros con Optuna ayudaron a explotar bien estas señales, pero el salto real vino de cómo representamos el comportamiento histórico.

En contraste, la decisión técnica que más nos limitó fue mantener una política de umbral/Top-N esencialmente global y estática para transformar los scores del modelo en recomendaciones concretas. Este enfoque provocó un sistema demasiado “abierto” en el Batch 1 (mucho recall, baja precisión) y luego demasiado conservador en el Batch 2 (precisión alta, recall muy bajo), sin una recalibración fina del umbral por semana, segmento o nivel de actividad del cliente. Incluso después del reentrenamiento, el modelo mejoró, pero la capa de decisión quedó por detrás en sofisticación: probablemente podríamos haber ganado puntos adicionales de F1 con una optimización más explícita de umbral y top-k por cliente alineada a la métrica de evaluación.

##### 6. ¿Qué hiperparámetro fue el más importante para su modelo?

Para ajustar el modelo final utilizamos Optuna sobre un `LGBMClassifier` dentro del pipeline de sklearn, optimizando PR-AUC en VALID, coherente con el foco del negocio en priorizar un ranking rico en verdaderos positivos. Entre los hiperparámetros explorados, el más influyente fue `num_leaves`, en combinación con `max_depth`, porque controla de forma directa la capacidad efectiva de los árboles para capturar interacciones no lineales entre recencia (`cp_recency`), tasas históricas (`prod_prior_rate`, `cust_cat_prior_rate`), actividad reciente (`cp_count_4w`, `cp_count_12w`) y atributos categóricos (tipo de cliente, región, marca). Valores bajos de `num_leaves` producían un modelo demasiado rígido, incapaz de explotar bien estas combinaciones, mientras que valores muy altos generaban sobreajuste en segmentos con pocos datos. La configuración óptima encontrada (`num_leaves ≈ 110`, `max_depth ≈ 12`, junto con un `min_child_samples` relativamente alto) logró un equilibrio adecuado entre capacidad y regularización, traduciéndose en un aumento consistente de PR-AUC (≈0.962 en VALID, ≈0.965 en TEST) y una mejora de F1 de alrededor de 1–2 puntos porcentuales frente a configuraciones alternativas.

##### 7. ¿Qué variable fue más influyente en las predicciones? ¿Cómo lo interpretan? ¿Cómo cambió su importancia con respecto a las otras variables a lo largo de los batch de datos?

Los valores SHAP del modelo LightGBM indican que la variable más influyente fue `customer_type_nan`, que actúa como una segmentación gruesa del tipo de cliente y aparece sistemáticamente con el mayor valor medio de |SHAP|. Inmediatamente después se sitúan las variables de historial y recencia del par cliente–producto (`cp_recency`, `prod_prior_rate`, `cp_prev_bought`, `cp_count_4w`/`cp_count_12w`), que son las que realmente ordenan la probabilidad de recompra: cuanto más reciente y frecuente es la relación cliente–producto, mayor es el empuje positivo sobre la predicción. A lo largo de los distintos batches, `customer_type_nan` se mantiene como separador global pero su peso relativo disminuye ligeramente, mientras que las variables de historial ganan importancia a medida que se enriquece el registro temporal. Variables categóricas más específicas, como `region_id_80`, `brand_BRAND 34` o `cust_cat_prior_rate`, tienen un rol más bien de refinamiento: ajustan la probabilidad dentro de un patrón ya marcado por recencia y frecuencia, sin desplazar a estas últimas del núcleo de importancia del modelo.

##### 8. ¿Qué aprendieron sobre el negocio a partir de los resultados del modelo?

A partir de los resultados del modelo, lo primero que se confirma es que el negocio de SodAI Drinks está fuertemente guiado por el historial reciente de compra. Variables como `cp_recency`, `cp_prev_bought`, `cp_count_4w`/`cp_count_12w` y `prod_prior_rate` muestran de forma consistente que lo que un cliente compró hace poco tiene mucha más probabilidad de ser recomprado que un producto “antiguo” en su historial. En términos comerciales, esto sugiere que la fuerza de ventas debería priorizar sistemáticamente, en cada llamada, los productos que forman parte del hábito reciente del cliente antes que empujar un catálogo amplio de alternativas nuevas.

También se observa que la segmentación importa: el tipo de cliente (`customer_type`) y la zona (`region_id`) generan patrones de respuesta distintos, lo que indica que hay canales y regiones donde el modelo es más confiable que en otros. Esto abre la puerta a estrategias diferenciadas, por ejemplo usando listas más largas y exploración en clientes con historial denso y buena respuesta histórica, y listas más acotadas y conservadoras en segmentos más ruidosos o menos activos. Además, las métricas muestran un sistema que opera en modo “alto recall, precisión moderada”: captura buena parte de las compras reales a costa de algunas recomendaciones que no se concretan, algo manejable si se controla el top-k por cliente para no sobrecargar al call center.

Por último, el hecho de que las mismas familias de variables (recencia, frecuencia, tipo de cliente, algunas categorías/marcas) se mantengan relevantes a lo largo de los cuatro batches sugiere que el comportamiento de compra es relativamente estable: no hay cambios bruscos en “qué explica” las ventas, sino ajustes suaves. Esto es una buena noticia para el negocio, porque permite construir políticas de recomendación estables sobre el modelo, afinando básicamente el volumen y la agresividad de las sugerencias según canal y respuesta observada, más que rediseñar la lógica comercial cada semana.

##### 9. ¿Qué limitaciones detectaron en su modelo o en los datos?

Detectamos varias limitaciones tanto en el modelo como en los datos, que ayudan a explicar por qué el rendimiento “real” en los batches es más modesto que el sugerido por las métricas de la Entrega 1. Primero, el sistema tiende a generar un volumen alto de recomendaciones por semana (miles de pares cliente–producto), lo que produce un F1 razonable pero con precisión solo moderada y muchos falsos positivos. En términos de negocio, esto implica una carga operacional elevada para el call center y para los clientes, porque se ofrecen muchos productos que finalmente no se compran. Esta situación no depende solo del modelo, sino también de la política de decisión: usamos un umbral global y un top-N casi uniforme por cliente, sin adaptar la intensidad de recomendación al tipo de cliente, canal o zona.

Además, el modelo fue entrenado sobre un universo recortado de pares cliente–producto, con una proporción de positivos bastante más alta que en el espacio total de candidatos. En la práctica hay un submuestreo implícito de negativos: no consideramos todos los “no compró este producto” posibles. Eso hace que las probabilidades tiendan a estar infladas y la calibración sea optimista: los umbrales que funcionaban bien en validación offline producen demasiadas recomendaciones y más falsos positivos cuando se evalúan en el esquema de competencia (conjuntos de pares por semana). De ahí la brecha entre AUC/PR-AUC cercanas a 0.97 en la Entrega 1 y F1 ≈ 0.40–0.49 cuando medimos en los batches semanales.

También hay limitaciones en la estrategia de evaluación. La partición temporal original train/valid/test es razonable para el modelamiento, pero la evaluación se hizo sobre el dataset tabular, no sobre el esquema set-based de recomendación. Eso tiende a sobreestimar el rendimiento en comparación con el escenario real, donde solo podemos recomendar unos pocos productos por cliente. En los batches, además, no observamos verdaderos TN, de modo que el classification_report estándar es poco informativo para la clase 0 y la métrica realmente útil es el F1 que calculamos a partir de los conjuntos de pares. Un esquema más realista requeriría trabajar explícitamente con métricas tipo precision@k, recall@k y validación temporal en ventanas rodantes.

Por último, la interpretabilidad mostró que el modelo depende fuertemente de unas pocas variables como `customer_type_nan` y `region_id_80`, lo que sugiere desbalances y patrones de missingness estructurados que pueden estar introduciendo artefactos. Al mismo tiempo, carecemos de señales clave de negocio como precios, promociones o stock, y la temporalidad se resume solo mediante agregaciones, sin explotar dinámicas secuenciales más ricas. En conjunto, el modelo explota bien la señal fuerte de recencia e historial, pero la combinación de submuestreo, evaluación optimista y una política de decisión global provoca un sistema que recomienda más de lo que el equipo comercial puede aprovechar eficientemente. Un diseño futuro debería integrar una evaluación alineada con producción, mejor calibración y umbrales/top-k diferenciados por segmento para controlar mejor el compromiso entre cobertura y relevancia.

### 📌 Conclusiones y Aprendizajes [1.0 puntos]

Este proyecto permitió recorrer de forma completa el ciclo de vida de un sistema de recomendación predictivo, desde la construcción del dataset hasta el despliegue productivo y su evaluación en condiciones cercanas al mundo real. A nivel de modelamiento, el trabajo mostró que un enfoque basado en un modelo de boosting, combinado con un buen diseño de atributos de recencia, frecuencia y afinidad cliente–producto, es capaz de capturar patrones relevantes del negocio y alcanzar métricas altas en validación offline. La optimización de hiperparámetros con Optuna y el análisis de interpretabilidad con SHAP ayudaron a entender por qué el modelo funciona y qué señales utiliza realmente para decidir, evitando que el proceso se reduzca a una caja negra.

Desde la perspectiva de MLOps, el diseño del pipeline con Airflow, el tracking sistemático en MLflow y la incorporación de lógica de reentrenamiento guiada por *data drift* demostraron la diferencia entre entrenar un modelo y operar un sistema de predicción. La experimentación semanal con batches nuevos evidenció que las métricas offline pueden ser optimistas, que el desempeño efectivo depende tanto del modelo como de la política de decisión (umbral, top-k por cliente) y que la adaptación temporal mediante reentrenamiento periódico es necesaria para mantener la calidad en el tiempo.

Finalmente, el análisis de resultados permitió extraer aprendizajes de negocio y de diseño de sistemas. El modelo confirmó que las compras recientes y la historia específica cliente–producto son determinantes en el comportamiento de demanda, mientras que ciertas regiones, categorías y tipos de cliente modulan ese efecto. Al mismo tiempo, se hicieron visibles limitaciones importantes: volumen elevado de recomendaciones, calibración perfectible y ausencia de algunas variables comerciales relevantes. En conjunto, el proyecto dejó como aprendizaje central que un buen sistema data-driven no depende solo de un modelo con buenas métricas, sino de la integración coherente entre datos, métricas, pipeline, monitoreo y objetivos operativos del negocio.

## 📁 Estructura de entrega

Se espera que su entrega cuente con la siguiente estructura de carpetas:

```bash
entrega_3/
│
├── informe.ipynb # Jupyter notebook con sus resultados
│
├── predictions/ # Carpeta con las predicciones para cada batch de datos
│
└── data/ # Carpeta con los datos realizados (lo que realmente ocurrió)
```

Mucho éxito!

<center>
<img src="https://gifdb.com/images/high/may-the-force-be-with-you-anakin-skywalker-n2g5o0pm4h6iylsx.gif" width="400" height="300">
